tests/unit/state_manager/test_step_state.py

```diff
@@ -0,0 +1,158 @@
+        state = SomeStepState(
+            part_properties={
+                "name": "foo",
+            },
+            project_options={
+                "number": 42,
+            },
+            files={"a"},
+            directories={"b"},
+        )
+        assert state.marshal() == {
+            "part-properties": {"name": "foo"},
+            "project-options": {"number": 42},
+            "files": {"a"},
+            "directories": {"b"},
+        }
+
+    def test_ignore_additional_data(self):
+        state = SomeStepState(extra="something")
+        assert state.marshal() == {
+            "part-properties": {},
+            "project-options": {},
+            "files": set(),
+            "directories": set(),
+        }
+
+
+class TestStepStatePersist:
+    """Verify writing StepState to file."""
+
+    def test_write(self, new_dir):
+        state = SomeStepState(
+            part_properties={
+                "name": "foo",
+            },
+            project_options={
+                "number": 42,
+            },
+            files={"a"},
+            directories={"b"},
+        )
+
+        state.write(Path("state"))
+        with open("state") as f:
+            content = f.read()
+
+        new_state = yaml.safe_load(content)
+        assert new_state == state.marshal()
+
+
+class TestStateChanges:
+    """Verify state comparison methods."""
+
+    def test_property_changes(self):
+        state = SomeStepState(
+            part_properties={
+                "name": "alice",
+                "pet": "spider",
+            },
+        )
+
+        # relevant properties didn't change
+        assert (
+            state.diff_properties_of_interest({"name": "alice", "pet": "eel"}) == set()
+        )
+
+        # relevant properties changed
+        assert state.diff_properties_of_interest({"name": "bob"}) == {"name"}
+
+    def test_project_options_changes(self):
+        state = SomeStepState(
+            project_options={
+                "number": 42,
+                "useful": False,
+            },
+        )
+
+        # relevant project options didn't change
+        assert (
+            state.diff_project_options_of_interest({"number": 42, "useful": True})
+            == set()
+        )
+
+        # relevant project options changed
+        assert state.diff_project_options_of_interest({"number": 50}) == {"number"}
+
+
+class TestHelpers:
+    """Tests for helper functions."""
+
+    @pytest.mark.parametrize(
+        "d1,d2,result",
+        [
+            [{}, {}, set()],
+            [{"a": 1}, {}, {"a"}],
+            [{}, {"b": 2}, {"b"}],
+            [{"a": 1}, {"b": 2}, {"a", "b"}],
+            [{"a": None}, {}, set()],
+            [{}, {"b": None}, set()],
+            [{"a": None}, {"b": None}, set()],
+            [{"a": 1}, {"a": 1}, set()],
+            [{"a": None}, {"a": 1}, {"a"}],
+            [{"a": 1}, {"a": 1, "b": 2}, {"b"}],
+            [{"a": 1, "b": 2}, {"a": 1}, {"b"}],
+        ],
```

Comment:
Missing at least the case of same key with different not-None values.

Reply:
Adding this test.